examples/recording_service/leveldb-integration/cpp/LevelDbReader.cxx

```diff
@@ -0,0 +1,637 @@
+         * UserDataValue, tailored to provide the validity of the sample or
+         * not. We won't attempt to load the dynamic data sample from the CDR
+         * buffer if the valid_data flag is not set
+         */
+        UserDataValue current_value = slice_to_user_type<UserDataValue>(
+                db_iterator_->value(),
+                value_buffer_);
+        std::shared_ptr<DynamicData> sample(
+                std::make_shared<DynamicData>(type_));
+        if (current_value.valid_data()) {
+            rti::core::xtypes::from_cdr_buffer(
+                    *(sample.get()),
+                    current_value.data_blob());
+        }
+        loaned_samples_.resize(num_samples_read + 1);
+        loaned_samples_[num_samples_read] = sample;
+        /*
+         * The read operation needs to return a collection of sample info
+         * objects too. Replay and Converter are fine with just returning the
+         * valid_data flag and the reception timestamp
+         */
+        std::shared_ptr<dds::sub::SampleInfo> sample_info(
+                std::make_shared<dds::sub::SampleInfo>());
+        DDS_SampleInfo native_sample_info = DDS_SAMPLEINFO_DEFAULT;
+        dds::core::Time reception_time = timestamp_to_time(current_timestamp_);
+        native_sample_info.reception_timestamp.sec =
+                (DDS_Long) reception_time.sec();
+        native_sample_info.reception_timestamp.nanosec =
+                (DDS_UnsignedLong) reception_time.nanosec();
+        native_sample_info.valid_data = current_value.valid_data()
+                ? DDS_BOOLEAN_TRUE
+                : DDS_BOOLEAN_FALSE;
+        (*sample_info)->native(native_sample_info);
+        loaned_infos_.resize(num_samples_read + 1);
+        loaned_infos_[num_samples_read] = sample_info;
+        num_samples_read++;
+        /* Advance to next DB position */
+        db_iterator_->Next();
+        if (db_iterator_->Valid()) {
+            current_key = slice_to_user_type<UserDataKey>(
+                    db_iterator_->key(),
+                    key_buffer_);
+            current_timestamp_ = current_key.reception_timestamp();
+            /*
+             * Check if the sample is out of the stream's time range; if that is
+             * the case, we should set the stream as finished
+             */
+            if (current_timestamp_ > end_timestamp_) {
+                next_sample_valid = false;
+            } else {
+                next_sample_valid = true;
+            }
+        } else {
+            next_sample_valid = false;
+        }
+    }
+    /*
+     * Fill out the collections passed as parameters by the service. We will use
+     * the internal vectors of samples and sample infos to get pointers
+     * to be stored in these collections
+     */
+    sample_seq.resize(num_samples_read);
+    info_seq.resize(num_samples_read);
+    for (int32_t i = 0; i < num_samples_read; i++) {
+        sample_seq[i] = loaned_samples_[i].get();
```

Comment:
I would indicate this works because Replay guarantees that for each read() and return_loan() follows, and hence we can use a single cached pool of samples

Reply:
Done